### DISCUSSION TITLE
RAG does not free VRAM

### DISCUSSION THREAD

**Discussion by treepoint** [open-webui/open-webui#6381] [General]
# Bug Report

## Installation Method

Docker, WSL, CUDA image.

## Environment

- **Open WebUI Version:** [e.g., v0.3.32]

- **Operating System:** [e.g., Windows 10]

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

When RAG stop processed question than its free VRAM. Technically — unloading RAG models for next use to give VRAM for another models. And won't load these models till it actually needed.

## Actual Behavior:
OpenWebUI eat VRAM on start, then eat another portion VRAM on update models, then eat another VRAM when make RAG request. And never give it back :)

## Description

**Bug Summary:**
RAG not free VRAM after process query and after update embedded/reranking models too.

## Reproduction Details
1. Set embedding model. My option — «ai-forever/ru-en-RoSBERTa».
2. Set reranking model. My option — «bge-reranker-v2-m3».
3. Click «save».
4. See minus 1.5 Gb VRAM.
5. Click «save» again, minus another 1.5 Gb VRAM.
6. Click till memory will fill totally.

And this VRAM will not be free in the near future. But if restart container it will be free again.

Similar things happen when i make RAG questions: first use — minus 1.5 Gb, second — minus another 1.5 Gb and so on.

And on start i see that it takes 1.5 Gb VRAM for, i think, the first init of models.

**Comment by dojoca** [comment]
Just bumping this... surely a variable can be set that would unload the RAG models from memory after a certain amount of time? At the moment the only way to do this is to restart the container.

**Comment by mahenning** [comment]
I can confirm for re-saving the model, I still have to test for using RAG.. for now I thought that using RAG does not increase VRAM usage. For loading/re-saving I have a PR almost ready that frees VRAM and also does not load the reranker if hybrid search is off. Unloading the model after a certain amount of no-usage time would require a bigger code re-write. As of right now, the model is loaded in the beginning and then referenced, which is very easy.

**Reply by mahenning** [reply]
I did test using RAG and with `jinaai/jina-embeddings-v3` I had some steady growth in VRAM usage after each new question, but not with `Snowflake/snowflake-arctic-embed-l-v2.0`. There, it was only one ram growth after the first question. Looking into it..

**Comment by rgaricano** [comment]
I don't think that it's useful, embedding it's an important function and embedd models are "small".

But i left this For Reference,

## Current Model Storage

The embedding model is currently loaded and stored in the FastAPI application state `app.state.ef` and the embedding function in `app.state.EMBEDDING_FUNCTION` . The model remains in memory throughout the application lifecycle with no explicit unloading mechanism.

## Implementation Approach

To implement an unload function, you would need to:

### 1. Create an Unload Endpoint

Add a new endpoint in the retrieval router similar to how the embedding configuration is updated [[1]](https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/src/lib/components/admin/Settings/Documents.svelte#L104-L125) :

```python
@router.post("/embedding/unload")
async def unload_embedding_model(request: Request, user=Depends(get_admin_user)):
    try:
        # Clear the embedding function and model from app state
        if hasattr(request.app.state, 'EMBEDDING_FUNCTION'):
            request.app.state.EMBEDDING_FUNCTION = None
        if hasattr(request.app.state, 'ef'):
            del request.app.state.ef
            request.app.state.ef = None
        
        # Force garbage collection
        import gc
        gc.collect()
        
        return {"status": True, "message": "Embedding model unloaded successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
```

### 2. Frontend Integration

Add an unload button to the Documents settings interface [[2]](https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/src/lib/components/admin/Settings/Documents.svelte#L851-L877) , similar to the existing update button:

```javascript
const unloadEmbeddingModelHandler = async () => {
    updateEmbeddingModelLoading = true;
    const res = await fetch(`${WEBUI_API_BASE_URL}/retrieval/embedding/unload`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${localStorage.token}`,
            'Content-Type': 'application/json'
        }
    });
    
    if (res.ok) {
        toast.success('Embedding model unloaded successfully');
    } else {
        toast.error('Failed to unload embedding model');
    }
    updateEmbeddingModelLoading = false;
};
```

### 3. Model Reloading Logic

You'd also need to modify the embedding function getter to handle the case where the model has been unloaded and needs to be reloaded on demand `retrieval.py:72-79`.

To implement on-demand reloading, you need to modify the `get_embedding_function` in `backend/open_webui/retrieval/utils.py` to check if the model is loaded and reload it if necessary:

### 3.1. Modify the Embedding Function Wrapper

The current `get_embedding_function` `utils.py:401-445 ` returns a lambda that directly uses the embedding function. You'd need to wrap this with lazy loading logic:

```python
def get_embedding_function_with_lazy_loading(
    embedding_engine,
    embedding_model,
    embedding_function,
    url,
    key,
    embedding_batch_size,
    azure_api_version=None,
    app_state=None,  # Pass app.state for reloading
):
    def lazy_embedding_function(query, prefix=None, user=None):
        # Check if model is unloaded and reload if necessary
        if embedding_engine == "" and (not hasattr(app_state, 'ef') or app_state.ef is None):
            # Reload the model
            app_state.ef = get_ef(
                embedding_engine,
                embedding_model,
                auto_update=False
            )
        
        # Use the original logic but with dynamic model checking
        if embedding_engine == "":
            return app_state.ef.encode(
                query, **({"prompt": prefix} if prefix else {})
            ).tolist()
        # ... rest of the original logic for other engines
    
    return lazy_embedding_function
```

### 3.2. Update the Model Update Handler

The existing `update_embedding_config` endpoint `retrieval.py:324-357`  already reloads the model when configuration changes. You can use similar logic for on-demand reloading.

### 3.3. Modify Query Processing

The embedding function is used in query processing `utils.py:325-398`  where it's passed to functions like `query_collection_with_hybrid_search`. The lazy loading wrapper would ensure the model is available when needed.

### 3.4. Handle Different Embedding Engines

For SentenceTransformers (default engine), the reloading logic uses `get_ef` `retrieval.py:118-138` . For external APIs (OpenAI/Ollama), no model reloading is needed since they're stateless HTTP calls.

## Integration Points

The reloading logic needs to be integrated at the application level where `app.state.EMBEDDING_FUNCTION` is used, particularly in:

1. **Document processing**: When files are processed for RAG `utils.py:522-567` 
2. **Embeddings API**: The generic embeddings endpoint `main.py:1343-1366`
3. **Query operations**: During vector similarity search operations

**Reply by mahenning** [reply]
My initial idea was a classwrapper that handles loading/unloading itself. When another function calls for embed, the class (instance) checks if it's unloaded, loads, then returns the embeddings. If it's already loaded, just return the embeddings. But yes, it would be a bit of work.
Still, the old model not correctly unloading when switching models should be fixed. It does not happen often in most cases, but with tight VRAM configs could lead to crashes.

**Reply by rgaricano** [reply]
The current update_embedding_config function is in https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/backend/open_webui/routers/retrieval.py#L324-L357

### Approach to cleanup logic before creating the new embedding function:
Before line 324, add cleanup logic for the existing embedding function:  
```
if request.app.state.ef is not None:  
    # Cleanup existing embedding function  
    if hasattr(request.app.state.ef, 'cleanup'):  
        request.app.state.ef.cleanup()  
    request.app.state.ef = None  
  
if request.app.state.EMBEDDING_FUNCTION is not None:  
    # Clear embedding function  
    request.app.state.EMBEDDING_FUNCTION = None  
  
# Force garbage collection  
import gc  
gc.collect()  
  
# For GPU models, clear CUDA cache  
if torch.cuda.is_available():  
    torch.cuda.empty_cache()
```

### Ollama-Specific Unloading
For Ollama models specifically, we can leverage the existing unload functionality. The system already has Ollama API integration `main.py:107-109` , so we can send unload requests to the configured Ollama instances before switching models.

**Comment by mahenning** [comment]
If you want to free VRAM after a certain amount of time, the best way is to use external embedding/reranking models and deploy them either via Ollama (which has inbuild timeout of models) or Llama.cpp with Llama-swap (which brings auto unloading to Llama.cpp). I don't think this feature will come to OWUI soon.

**Comment by mahenning** [comment]
I've added a draft PR to fix the VRAM issue for reloading the models (https://github.com/open-webui/open-webui/pull/16779). Code review and feedback is welcome!

**Comment by ldpg-dev** [comment]
@mahenning I use an external reranker, external embedding and external web search. But every time I do a web search, OWUI loads all indexes in RAM and never cleans them up even though the web search data is never used again. It can easily be cleaned up... This happens regardless of the Vector DB and after a few web searches it has already taken 6gb of RAM, then another and another... Which is problematic for long running containers.

This is to say OWUI has a widespread memory management issue not only for VRAM but also for RAM.

**Reply by rgaricano** [reply]
this should be managed for each db engine,

For reference,

e.g of unload collection function to execute after searched collectior was saved to db (e.g. add call in backend/open_webui/routers/retrieval.py#L2016),
for chromadb:
```
def unload_collection(self, collection_name: str):  
    """Unload collection from memory without deleting it from disk"""  
    try:  
        if self.has_collection(collection_name):  
            # ChromaDB doesn't have explicit unload, but we can force garbage collection  
            # by removing the collection reference and triggering cleanup  
            collection = self.client.get_collection(name=collection_name)  
            del collection  
              
            # For HTTP client, we could also reset the connection  
            if hasattr(self.client, '_session'):  
                self.client._session.close()  
                  
            log.info(f"Unloaded collection {collection_name} from memory")  
            return True  
    except Exception as e:  
        log.debug(f"Error unloading collection {collection_name}: {e}")  
        return False
```
for pgvector:
```
def unload_collection(self, collection_name: str):  
    """Unload collection from memory by closing session and clearing cache"""  
    try:  
        if self.has_collection(collection_name):  
            # Close current session to free memory  
            self.session.close()  
              
            # Recreate session for future operations  
            if hasattr(self, 'SessionLocal'):  
                self.session = scoped_session(self.SessionLocal)  
              
            log.info(f"Unloaded collection {collection_name} from memory")  
            return True  
    except Exception as e:  
        log.debug(f"Error unloading collection {collection_name}: {e}")  
        return False
```

a more simple solution for no multiuser enviroment is close the db connection and reopen after all web searches are done, it should work for any db engine.
e.g.
```
# In process_web_search() after the final return statement  
try:  
    await run_in_threadpool(  
        save_docs_to_vector_db,  
        request,  
        docs,  
        collection_name,  
        overwrite=True,  
        user=user,  
    )  
      
    # Reinitialize vector DB client after ALL web searches are done  
    from open_webui.retrieval.vector.factory import Vector  
    from open_webui.config import VECTOR_DB  
    global VECTOR_DB_CLIENT  
      
    if hasattr(VECTOR_DB_CLIENT, 'close'):  
        VECTOR_DB_CLIENT.close()  
      
    VECTOR_DB_CLIENT = Vector.get_vector(VECTOR_DB)  
    log.info(f"Reinitialized vector DB client after completing all web searches for collection: {collection_name}")  
      
    return {  
        "status": True,  
        "collection_names": [collection_name],  
        "filenames": urls,  
        "loaded_count": len(docs),  
    }  
except Exception as e:  
    log.debug(f"error saving docs: {e}")
```

**Reply by mahenning** [reply]
I did not work with web search yet, but I'll look into it. Maybe I'll find something. Do you use open-webui internal chromadb as vector storage? 
> It can easily be cleaned up

How do you do that? The more information I get the better I can look into it.

**Reply by rgaricano** [reply]
right now, I'm using Postgresql (but I can easyly change to test in chromadb, I have it as lab)

**Reply by ldpg-dev** [reply]
@mahenning I have a job that runs every week just to cleanup orphaned collections that have the web search prefix. This cleans up the indexes and frees up RAM by triggering garbage collection in the Vector DB container and somehow it seems to also clean up in Open WebUI's container after a few hours... This should be managed internally by just cleaning up after each web search, you can clean up both the Vector DB after search as @rgaricano is proposing, and you can clean up the actual in-memory data after a web search. That alone should be good enough but perhaps a more refined solution can be better since memory management seems to be an issue in some other places too.

**Comment by TheSpaceGod** [comment]
Hi @mahenning,

Do you think the following (AI assisted) purposal would fit into the added logic you added in your PR?


***

**Feature Request: Option (ENV Variable) to unload RAG models from VRAM after each use**

### The Problem

The RAG implementation in Open WebUI is a powerful feature. However, on VRAM-constrained systems (like most consumer graphics cards: 8GB, 12GB, 16GB, etc.), it comes with a significant trade-off.

When a RAG-enabled query is made, Open WebUI loads the embedding and/or reranking models into VRAM. For performance, these models are then cached and remain in VRAM for the next query.

While this caching is excellent for speed, it permanently "reserves" a large chunk of VRAM—often 2-4 GB or more, depending on the models selected. This reserved VRAM is then *unavailable* for the main generator model being served by `ollama`.

### The Impact on VRAM-Limited Users

For users on consumer GPUs, VRAM is the single most valuable resource. The 2-4GB of VRAM held by the RAG models could be the difference between:
* Running a 7B model vs. a 13B model.
* Running a heavily quantized model vs. a higher-quality, less-quantized version.
* Using a small context window vs. a much larger one.
* Running a query successfully vs. getting an "Out of Memory" (OOM) error.

In many user workflows, the RAG step is sequential. The user is willing to accept a *slower* RAG step if it means they can use a *better* generator LLM for the final answer.

### Proposed Solution

I propose adding an **optional** setting, ideally as an environment variable (e.g., `RAG_UNLOAD_AFTER_USE=true`), to change this behavior.

When this variable is set, Open WebUI would follow this "serialization" workflow for each RAG-enabled query:

1.  **Load:** Load the embedding/reranking models into VRAM.
2.  **Execute:** Perform the embedding and reranking steps to retrieve the context.
3.  **Unload:** Immediately unload the embedding/reranking models from VRAM, freeing the memory.
4.  **Generate:** Pass the context to the `ollama` LLM, which can now use the maximum VRAM available on the card.

### Why This Is Useful

This would allow users to "time-share" their VRAM. We accept that this will make RAG queries slower, as the models must be reloaded for each query. However, this is a trade-off many of us would *gladly* make (especially if this data is still being cached by file systems like ZFS via system RAM or NVMe based caching).

This feature would give users with limited VRAM the power to run high-quality RAG pipelines *and* large, capable `ollama` models, making Open WebUI even more versatile for the enthusiast community.

Thank you for considering this!

**Reply by rgaricano** [reply]
I was doing some test to do this, with a lazy load for the embedding function....
For that function work propertly embedd models have to be external, if is the default SentenceTransformers the load/unload can be a problem instead that a solution.

Instead of do that, finally, I adopted another solution: load the main ollama model before launch open-webui, then the main model always is 100% loaded in GPU & open-webui embedd use the rest & cpu.
If I start ollama and open-webui, when ollama load the model the main model is loaded 16%CPU/84%GPU...slowering the generation.
For do that I created a oneshot service that requiere&after ollama for load the model (and required for open-webui service:
```
#service for load a ollama model before start open-webui
[Unit]  
Description=Preload Ollama Model  
After=0-ollama.service  
Requires=0-ollama.service  
  
[Service]  
Type=oneshot  
ExecStart=/usr/local/bin/ollama run cogito:14b ""  
User=ollama  
Group=ollama  
Environment="PATH=/usr/local/bin:/usr/bin:/bin" 
  
[Install]  
WantedBy=multi-user.target
```
and in open-webui service:
```
[Unit]
Description=Openwebui-DEV Service
After=network-online.target
Requires=0-ollama.oneshot_run.service
...
```

**Reply by TheSpaceGod** [reply]
While the approach you suggest prioritizes the Ollama above all else (which is probably the most important thing), it still does not implement proper constrained resource sharing that would occur if each element was loaded/unloaded sequentially (ensuring each step gets max possible resources).  I feel something like this still needs to be implemented for local resource constrained environments running on consumer GPUs.

**Reply by rgaricano** [reply]
I'm working on it (for default sentenceTransformers), with good results, now I can upload hundred of files without any oom.
Basicly I implemented batch processing...and other minor changes (as check for files permited extensions before upload its, link  files to collections when its are duplicated,...)
